File: medium/Surrounded_Regions.py
class Solution:
    def __init__(self):
        self.width = 0
        self.length = 0
        self.queue = []
        self.safe = []
        self.safe_key = False
        
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        self.length = len(board)
        self.width = len(board[0])
        def check(i, j):
            self.queue.append((i, j))
            if i != 0:
                if board[i-1][j] == 'O' and (i-1, j) not in self.queue:
                    check(i-1, j)
            if j != 0:
                if board[i][j-1] == 'O' and (i, j-1) not in self.queue:
                    check(i, j-1)
            if i != self.length-1:
                if board[i+1][j] == 'O' and (i+1, j) not in self.queue:
                    check(i+1, j)
            if j != self.width-1:
                if board[i][j+1] == 'O' and (i, j+1) not in self.queue:
                    check(i, j+1)
                
        for i in [0, self.length-1]:
            for j in range(self.width):
                if board[i][j] != 'X':
                    check(i, j)
                    for l, w in self.queue:
                        board[l][w] = '*'
                    self.queue = []
                    
        for j in [0, self.width-1]:
            for i in range(self.length):
                if board[i][j] != 'X':
                    check(i, j)
                    for l, w in self.queue:
                        board[l][w] = '*'
                    self.queue = []
        
        for i in range(self.length):
            for j in range(self.width):
                if board[i][j] == 'O':
                    board[i][j] = 'X'
                if board[i][j] == '*':
                    board[i][j] = 'O'
        return board

File: cpp/medium/Surrounded_Regions.cpp
class Solution {
public:
    void solve(vector<vector<char>>& board) {
        int n = board.size();
        int m = board[0].size();
        vector<vector<int>> connect_edge;
        int j, i;
        for (int j = 0; j < n; ++j){
            for (int i = 0; i < m; ++i) {
                if (board[j][i] == 'O'){
                    if (j == 0 || i == 0 || j == n-1 || i == m-1) connect_edge.push_back(vector<int> {j, i});
                    else board[j][i] = '1';
                }
            }
        }

        for (int idx = 0; idx < connect_edge.size(); ++idx) {
            j = connect_edge[idx][0];
            i = connect_edge[idx][1];
            if (i > 0 && connect(board, j, i-1)) connect_edge.push_back(vector<int> {j, i-1});
            if (j > 0 && connect(board, j-1, i)) connect_edge.push_back(vector<int> {j-1, i});
            if (i < m-1 && connect(board, j, i+1)) connect_edge.push_back(vector<int> {j, i+1});
            if (j < n-1 && connect(board, j+1, i)) connect_edge.push_back(vector<int> {j+1, i});
        }

        for (int j = 0; j < n; ++j){
            for (int i = 0; i < m; ++i) {
                if (board[j][i] == '1') board[j][i] = 'X';
            }
        }
    }
    
    bool connect(vector<vector<char>>& board, int j, int i) {
        if (board[j][i] == '1') {
            board[j][i] = 'O';
            return true;
        }
        else return false;
    }
};
